File: server/src/routes/explorer.ts
import { Router, Request, Response } from 'express';
import { serverModel } from '../models/server.js';
import { isAgentConnected, requestFromAgent } from '../websocket/agentRegistry.js';
import { logger } from '../utils/logger.js';

const router = Router();

// GET /api/servers/:id/explore?path=/
router.get('/:id/explore', async (req: Request, res: Response) => {
  try {
    const server = serverModel.findById(req.params.id);
    if (!server) return res.status(404).json({ error: 'Server not found' });

    const remotePath = (req.query.path as string) || '/';

    if (!isAgentConnected(server.id)) {
      return res.status(503).json({ error: 'Agent is not connected' });
    }

    // Request file listing from the agent via WebSocket
    const result = await requestFromAgent(server.id, {
      type: 'fs:browse',
      payload: { path: remotePath },
    }) as Record<string, unknown>;

    if (result.error) {
      return res.status(500).json({ error: result.error });
    }

    res.json(result.entries || []);
  } catch (err: any) {
    logger.error({ err, serverId: req.params.id, path: req.query.path }, 'Failed to explore remote path');

    const message = err instanceof Error ? err.message : 'Exploration failed';

    if (message.includes('Permission denied') || message.includes('EACCES')) {
      return res.status(403).json({ error: 'Permission denied', code: 'PERMISSION_DENIED', path: req.query.path });
    }

    if (message.includes('No such file') || message.includes('not found')) {
      return res.status(404).json({ error: 'Path not found', code: 'PATH_NOT_FOUND', path: req.query.path });
    }

    res.status(500).json({ error: message });
  }
});

export default router;

File: backup-server-rs/src/routes/explorer.rs
use crate::error::AppError;
use crate::models::server;
use crate::state::AppState;
use axum::extract::{Path, Query, State};
use axum::Json;
use serde::Deserialize;
use std::sync::Arc;

#[derive(Deserialize)]
pub struct ExploreQuery {
    pub path: Option<String>,
}

pub async fn explore(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Query(query): Query<ExploreQuery>,
) -> Result<Json<serde_json::Value>, AppError> {
    let remote_path = query.path.unwrap_or_else(|| "/".into());

    // Verify server exists
    let db = state.db.clone();
    let id2 = id.clone();
    let srv = tokio::task::spawn_blocking(move || {
        let conn = db.get()?;
        server::find_by_id(&conn, &id2)
    })
    .await
    .map_err(|e| anyhow::anyhow!(e))??;

    if srv.is_none() {
        return Err(AppError::NotFound("Server not found".into()));
    }

    if !state.agents.is_connected(&id) {
        return Err(AppError::ServiceUnavailable("Agent is not connected".into()));
    }

    let message = serde_json::json!({
        "type": "fs:browse",
        "payload": { "path": remote_path },
    });

    let result = state
        .agents
        .request_from_agent(&id, message, 30_000)
        .await
        .map_err(|e| {
            let msg = e.to_string();
            if msg.contains("Permission denied") || msg.contains("EACCES") {
                AppError::BadRequest(format!("Permission denied: {}", remote_path))
            } else if msg.contains("No such file") || msg.contains("not found") {
                AppError::NotFound(format!("Path not found: {}", remote_path))
            } else {
                AppError::Internal(e)
            }
        })?;

    if let Some(error) = result.get("error") {
        return Err(AppError::Internal(anyhow::anyhow!("{}", error)));
    }

    let entries = result.get("entries").cloned().unwrap_or(serde_json::json!([]));
    Ok(Json(entries))
}
